### DISCUSSION TITLE
CC Fest presentation?

### DISCUSSION THREAD

**Discussion by tabreturn** [py5coding/py5generator#28] [General]
It's Processing's 20th anniversary. The foundation is organising a [CC Fest](https://ccfest.rocks), looking for speakers on various Processing-related topics. If you think it's time for py5 to get some more exposure, somebody could pitch a presentation?

https://docs.google.com/forms/d/e/1FAIpQLSdfw58H9vU-2uZqYgh3kr3Bc37vEiX1FTFxDfMSDcpSbFQ_pg/viewform

**Comment by hx2A** [comment]
This is a great idea. Thank you for suggesting it. Where did you find out about this? What is the submission deadline? Is there more information about this event in August?

I will submit something, but note I've submitted things to past Processing events and have never gotten accepted. I'll do my best to change that this time around.

**Reply by tabreturn** [reply]
> Where did you find out about this? What is the submission deadline? Is there more information about this event in August?

Somebody from the Processing foundation sent me a link. I'm guessing they'll announce the details when they have speakers locked in. Glad to hear you'll submit something :+1:! I'm guessing you'll show off py5? If you don't, I'll try, haha.

Let me know if you'd like me to review your submission.

**Reply by hx2A** [reply]
> Let me know if you'd like me to review your submission.

Definitely, I will take you up on that offer. Most likely I'll do a presentation that gives an overview of the library and a demo for attendees via Jupyter Hub.

> I'm guessing you'll show off py5? If you don't, I'll try, haha.

Actually, if you are interested, I'd be more than happy if you submitted something also. I don't want py5 to become the "Jim-show" or be all about me. It would communicate that there is growing interest in py5 and there is an emerging community. This is an important goal.

**Reply by tabreturn** [reply]
> gives an overview of the library and a demo for attendees via Jupyter Hub.

I think that's a great angle to take. It exhibits the novel py5 contributions (Processing running in a notebook, with Python) and allows attendees to partake virtually.

I can't promise I'll submit anything. But if I can think of something sufficiently novel, I'll submit it.

**Reply by hx2A** [reply]
This past week was crazy, but at least I completed what I needed to complete to do another release.

I'll think about a presentation this week and get back to you.

> I can't promise I'll submit anything. But if I can think of something sufficiently novel, I'll submit it.

How about something related to py5 and Thonny?

**Reply by tabreturn** [reply]
> How about something related to py5 and Thonny?

Cool. I'll see what I can put together.

EDIT: done!

**Reply by hx2A** [reply]
It doesn't look like this is going to happen, I'm afraid. Right now I'm behind the eight ball with work related tasks and can't possibly put together a presentation to deliver in a week and a half. There's only so much I can do. I do feel bad about this, because this project means a lot to me. Although there will be other opportunities later, so I'm not overly worried.

**Reply by tabreturn** [reply]
No worries, @hx2A. Your work on py5 is fantastic! As you've noted, there'll be other opportunities. If my proposal is accepted, I'll be flying the flag for py5 at CC Fest :flags:

**Reply by hx2A** [reply]
Thank you, I appreciate your support. Let me know if your proposal is accepted, as I'll want to do a release before then to fix #37 .

I just noticed I have a medical procedure I need to prepare for the Monday after that weekend...that's going to be a rough weekend.

**Reply by tabreturn** [reply]
Will do. Wishing you the best for the procedure :crossed_fingers:

**Reply by tabreturn** [reply]
@hx2A -- my proposal has been accepted :smile:. Seeing you can't present, I'll touch on py5's other superpowers, maybe show off the mybinder examples. 

It would be excellent if #37 is resolved by then. I've got a [plugin working for Thonny](https://github.com/tabreturn/thonny-py5mode) (on Linux, that should run on Windows with the #37 fix). I was familiarizing myself with the Thonny architecture, so there's plenty room for improvement, but I plan to refine it further.

**Reply by hx2A** [reply]
@tabreturn fantastic! I'll do what I can to get #37 fixed and a new release out ASAP. I'll aim for Monday or Tuesday of next week.

**Reply by hx2A** [reply]
@tabreturn I have finished all of the changes I intend to make for the next release:

* 3 character web colors
* fix #37 
* use Processing 4.0b1

If you like you can test this using this branch: 

https://github.com/hx2A/py5/tree/0.5a2rc

Which you can install with something like `pip install git+https://github.com/hx2A/py5.git@0.5a2rc#egg=py5`.

If everything looks good I will do the release on Monday or Tuesday of next week.

If there's anything I can do to help you be more successful on the 22nd, please let me know.

**Reply by tabreturn** [reply]
Awesome work, @hx2A! I'm loving the 3-digit hex colors. If you manage to release this before CC Fest, that would be fantastic (any Windows Thonny users could install the Thonny plug-in, and I wouldn't have to change anything).

**Reply by hx2A** [reply]
@tabreturn I'm loving the 3 digit hex colors also. I've been using them all the time since I added the feature.

Everything look good to you on Windows? No problems running static sketches in Thonny?

I changed very little in this release so it is unlikely I am introducing any new problems. I am determined to release this before CC Fest.

**Comment by villares** [comment]
Hey people!

I have just tested @tabreturn's thonny-py5mode on a Windows machine and it works wonderfully 🥇 

I see no more reason for my tentative thonny-py5runner anymore I think I'll just delete it, I'm glad it inspired Tristan's sprint 😄 

On other news, I have sent a talk proposal for the Brazilian Pycon (it is called Python Brasil because journalists used to think the language was called Pycon...) and if it is selected I think I'll be able to show off some py5 😆

**Reply by hx2A** [reply]
@villares Great! Did you test with the py5 release candidate from github or did you install py5 from the pypi repository?

Awesome, I hope your proposal gets accepted!

**Reply by villares** [reply]
I assume it was the one from pypi. I can't remember installing it (maybe Tristan's plug-in did it, maybe I did it last week, it's not my personal computer), Thonny tells me it is 0.5a1.

**Comment by hx2A** [comment]
There is a big in that version where static mode sketches will not run
correctly on windows. The next release will fix that.

**Reply by tabreturn** [reply]
> There is a bug in that version where static mode sketches will not run correctly on windows

Yes, this is for the imported mode (*py5 > Run imported mode sketch*) that uses run_sketch.py -- which I can confirm is fixed in 0.5a2rc 🥳 

Thanks, @villares -- the plug-in gets the job done, but there's definitely room to improve it. I've learned a lot more about Thonny's code-base, and I have some plans to refine and improve things.

**Reply by villares** [reply]
imported mode seems fine:

![image](https://user-images.githubusercontent.com/3694604/129423388-aea5ff96-bf29-4c15-8d99-14c3048d6d20.png)

**Reply by villares** [reply]
I have to disclose this is an unconventional Thonny install:
I installed Python 3.9 first on Windows, then `pip install thonny`. So from terminal I type `python -m thonny` and I get Thonny 3.3.14

And it uses Python 3.9.6 for everything:
```
Python 3.9.6 (C:\Users\alexandre.villares\AppData\Local\Microsoft\WindowsApps\PythonSoftwareFoundation.Python.3.9_qbz5n2kfra8p0\python.exe)
```

![image](https://user-images.githubusercontent.com/3694604/129423567-4a7e9496-736f-400e-804b-92e9c55e0e52.png)

**Reply by hx2A** [reply]
Interesting...you'll have to ask @tabreturn if there is a way to get the release candidate version of py5 into your setup. That is the better version to test with.

**Comment by tabreturn** [comment]
To summarize this discussion: @tabreturn will present py5 at CC Fest (Processing 20th Anniversary edition) next weekend. The outline for the 60-minute presentation is looking something like this:

1. A very brief overview of Processing's Python history.
2. A brief, high-level overview of how Processing.py works and how py5 is different.
3. The presenter will introduce some py5 basics here using mybinder, contrasting a module mode and py5bot example (attendees might follow along here).
4. Move into Thonny at this point. Attendees might / might not be able to follow along (to wait on everyone to download/install/etc Thonny is impractical and not a good use of presentation time). The idea for this part is to show off py5 as a Processing.py/Python Mode alternative with Python 3 features, using some cool Python 3 libraries (maybe Pymunk).
5. If there is time, some closing Q & A.

The Thonny py5 plugin should work on Windows and Linux by then (see fix #37).

Let me know if there's anything I should change/remove/add.

**Reply by hx2A** [reply]
Will you mention p5.py? That is another Python version of Processing that has had several rounds of Google Summer of Code contributions.

Do you want more information about how py5 works? To be brief, I would just say it uses CPython and the Processing jars, and gets everything to work together using jpype.

In [another thread](https://github.com/hx2A/py5generator/discussions/12#discussioncomment-804014) I wrote about the goals of py5. Specifically:

> My goal is to create a new version of Processing that fits into the larger Python ecosystem. Python syntax is a byproduct of that, but it isn't the core goal. This also isn't an attempt to replace or compete with any other version of Processing. Py5 isn't and can't be a drop-in replacement for processing.py, even without the camel case / snake case differences. I made different design choices in pursuit of a different goal, resulting in different libraries that work differently.

A few specific things py5 can do that demonstrate how it fits into the larger Python ecosystem are:

* using numpy arrays to adjust pixels (np_pixels)
* converting matplotlib charts or PIL images to Py5Image objects (with convert_image)
* 2 Jupyter notebook kernels, plus the ability to run on mybinder
* use Python debugger
* built in support for line_profiler for performance tuning

I'd prefer for py5 to stand alongside the other versions of Processing and provide new opportunities for coders to explore what they can do with creative coding. Underneath the processing, processing.py, p5, processing.js, coffeescript processing, etc libraries are a set of core values the revolve around education, inclusiveness to the tech world, and creative exploration/expression, and I see py5 as a library that is collaborating with the other libraries to advance those core values.

**Reply by hx2A** [reply]
And I wrote all that to provide you with information to draw from as you put together your presentation. Don't feel like you have to include any part of it if it isn't a good fit for what you will talk about.

**Reply by tabreturn** [reply]
Thanks for that!

> Will you mention p5.py?

Yeah, I thought of mentioning p5.py among points 1 and 2, helping clarify all the Jython/JPype/etc. stuff that's happening under the hood of these different Python Processing alternatives. @villares' [comparison table](https://github.com/villares/Resources-for-teaching-programming) will, no doubt, feature here.

> that has had several rounds of Google Summer of Code contributions.

Just a thought: maybe some GSoC participation for py5 next round ...

> In another thread I wrote about the goals of py5

Thanks. I'm planning to highlight the goals of the project (both technically and philosophically). 

The proposal is titled "Thonny + py5: A Python 3 Environment for Processing", so I have to be careful not to drift too far off that topic. The proposal was targeted at Thonny + py5  (in case you submitted something).

Thanks for those topic suggestions. I'll see what fits nicely. I hadn't taken a proper look at `convert_image()` or even loading images into py5 yet -- this is exciting. Incidentally, I've used [pdb in Processing.py](https://discourse.processing.org/t/debugger-for-python-mode/9426)

**Reply by hx2A** [reply]
> Just a thought: maybe some GSoC participation for py5 next round ...

I have no idea how that process works, but if I could get py5 into the summer of docs program, I'd be very grateful.

> The proposal is titled "Thonny + py5: A Python 3 Environment for Processing", so I have to be careful not to drift too far off that topic.

Whatever you think is best! Whatever you don't include, I'm happy to have at least provided information that I'm sure will be helpful at a later date. There's a lot of information about py5 that I need to get out of my head and into some form that others can read and benefit from.

> I hadn't taken a proper look at convert_image() or even loading images into py5 yet -- this is exciting.

It is particularly valuable for converting SVG images to Py5Images. The Processing library does not support the full SVG specification, so it sometimes doesn't do so well wtih SVG images that it did not create itself. The convert_image() method will use Cairo to rasterize an SVG, and it consistently gives good results.

Another neat thing about convert_image() is the ability to extend its functionality with register_image_conversion(). If there is some graphical image format that someone uses a lot, they can register new functions for convert_image() to use. This will also be an easy way for people to contribute to py5 with PRs.

Also, don't forget about create_image_from_numpy().

**Reply by hx2A** [reply]
@tabreturn Good luck tomorrow! (Or today, depending on your time zone)

I still have that medical thing on Monday but expect to be able to watch part of cc fest tomorrow. If it doesn't make you uncomfortable, I'll probably be in the audience when you are presenting.

**Reply by tabreturn** [reply]
Thanks, @hx2A! I'm busy working on the presentation now. I hope all goes well with your procedure.

**Reply by hx2A** [reply]
@tabreturn, that was a great presentation! I even learned some things. I didn't know that you could use a debugger like that to pause a running program. I would have guessed there'd be some kind of timeout if the Java call to Python took too much time to complete.

Here is the link to the JupyterHub setup. I was trying to find it during the presentation but wasn't fast enough:

https://github.com/2i2c-org/peddie-image

They have a JupyterHub setup for their students. To run Processing and therefore py5 on a server like that, you need Java 11 and a display of some kind. The need for a display can be satisfied with Xvfb, a "virtual frame buffer" which is basically a fake monitor. Also, Python 3.8. 

Google Colab is on 3.6, but someday they will move to 3.8, and py5 will be runnable there.

What was the other platform the participant was asking about? Code something?

I really like your cheat sheet. That is a great idea. I did note there are a few camelCase issues in the typography and random sections. Also, py5 has a random_int function, which is easier to use than int(random()). You should also mention random_choice, which I use all the time.

Something that I thought of during your presentation is that Processing is going to move to Java 15 in the fall (I believe), and py5 will move with it. How will that affect Thonny plugin users? How does Thonny handle upgrades to libraries anyway? Since people need to use your menu option to download the JDK, there would have to be something to invalidate the downloaded JDK and download another one.

You were up at 4 am for this? You are dedicated!

**Reply by tabreturn** [reply]
Cheers, @hx2A! It's a pleasure to show off your work.

@mcintyrelabs was the person asking about the Jupyter stuff. Can't recall the other platform he mentioned -- maybe [Coding Rooms](https://www.codingrooms.com).

I rushed out the cheat sheet last minute :sweat_smile:; I've made some corrections now, thanks.

Yeah, the Thonny plug-in is very much a work-in-progress. Will have to look at how I might handle Processing/Java updates.

Good thing I don't have to go into work today ... I'm off to catch up on some sleep.

**Comment by sawula** [comment]
@tabreturn, i’m sorry i missed it.  best way to catch up?

**Reply by tabreturn** [reply]
https://github.com/tabreturn/cc-fest-py5

**Reply by tabreturn** [reply]
Another opportunity to present py5, maybe?

https://nightingaledvs.com/outlier-is-back-in-2022-apply-to-speak/

**Reply by hx2A** [reply]
@tabreturn Excellent. I will apply for this. Thank you. :)

**Reply by hx2A** [reply]
Is this in person or online? It is unclear from the website.

**Reply by tabreturn** [reply]
Hmm. Weird. I've no idea.

**Reply by hx2A** [reply]
Probably hybrid or online only. I can't imagine anyone would plan an in person event like that many months in the future.

BTW, how's it going with the lockdown in NZ?

**Reply by tabreturn** [reply]
Yeah, probably hybrid, I guess.

I'm outside of Auckland, so they're starting to relax things for us. It'll be a long time before any kind of international travel is practical.

**Reply by hx2A** [reply]
> Another opportunity to present py5, maybe?
> 
> https://nightingaledvs.com/outlier-is-back-in-2022-apply-to-speak/

@tabreturn I just submitted my application for this. I'll find out in November sometime.

Thanks again for the suggestion!

**Reply by tabreturn** [reply]
Nice one, @hx2A! I'm submitting something to Pycon (NZ) this week.

**Reply by tabreturn** [reply]
How about a py5 (and, maybe, Processing-Python) themed Feral File exhibition?
https://docs.google.com/document/d/1R5988iID938TH6bWirkG_AyI6liCufQdXr5pX6t9etk/edit

**Reply by hx2A** [reply]
> How about a py5 (and, maybe, Processing-Python) themed Feral File exhibition? https://docs.google.com/document/d/1R5988iID938TH6bWirkG_AyI6liCufQdXr5pX6t9etk/edit

This is an interesting idea. I had not heard of Feral File. As more people start using this and we start building a collection of py5 works, this could be a good way to promote the library and demonstrate that it can be used for artistic purposes.

**Reply by hx2A** [reply]
> > Another opportunity to present py5, maybe?
> > https://nightingaledvs.com/outlier-is-back-in-2022-apply-to-speak/
> 
> @tabreturn I just submitted my application for this. I'll find out in November sometime.
> 
> Thanks again for the suggestion!

@tabreturn Didn't get approved. Oh well, that's life. I'll try again somewhere else!

**Reply by tabreturn** [reply]
Bummer, @hx2A. Oh well, as you say, there will be other opportunities 🥲
